DESIGN.md

```diff
@@ -61,8 +61,90 @@ specific authorizations
 - capabilities: a request carries a token that contains a set of rights
 that will be used for authorization, instead of deploying ACLs on every node
 
+
+## Structure and semantics
+
+A biscuit is structured as a cryptographic, append-only list; its elements are
+called *caveats*, and describe authorization properties.  As with Macaroons,
+an operation must comply with all caveats in order to be allowed by the biscuit.
+
+Caveats describe which operations are authorized by providing predicates over
+the operation's attributes.
+
+Attributes are data, associated with the operation,
+that is known when the policy is evaluated, such as an identifier for the
+ressource being accessed, the type of the operation (read, write, append, ...),
+the operation's parameters (if any), the client's IP address or a
+channel-binding value (like the TLS transcript hash).
+
+Available attributes, and their type, are known ahead of time by the verifier.
+Some of those attributes are *critical*, and all caveats must provide a *bound*
```

Comment:
```suggestion
Some of those attributes are *critical*, and the first caveat must provide a *bound*
```

That's @tarcieri version.  I would rather enforce the property for all caveats because it means we guard against the same kind of mistake in attenuation too:
- the minting of new credentials may be done by a service which already has a caveat on its biscuit (for instance, an login system might get a biscuit with time-limited validity, but valid for all users, ressources, and operations);
- there are other contexts than credential minting (such as delegation) where accidental authority in attenuation is a major issue.

Reply:
adding bounds for all those attributes might make a token that is too large (let's remember, it should fit in a cookie), especially for the first caveat that might have the largest access. How could we make that as small as possible?

Reply:
@Geal How big would the representation of `foo: Any` be?  That is the max. overhead, per caveat and per-critical property, and we can likely shrink it down to a handful of bytes.

The main thing is that you *want* to always provide bounds for critical properties (cf. the discussion in Slack. TL;DR: unbounded critical properties confer accidental authority when the property is added or the property gains new possible values), so it's not so much of an overhead rather than preventing people from accidentally issuing tokens missing those caveats.

I guess that, if biscuit size becomes an issue, we could support a compression scheme in settings where preceeding caveats are readable (i.e. in a public-key-signed token). For example, the representation of a predicate could refer to predicates in preceeding caveats when they share a (sub)term.
I'm not super sure that it makes a ton of sense to argue about size when the representation format isn't nailed down, as we can't yet measure the overhead (in bytes).